cmd/blockchaincmd/change_weight.go

```diff
@@ -141,59 +134,133 @@ func setWeight(_ *cobra.Command, args []string) error {
+
+	if float64(vdrInfo.Weight) > allowedChange {
+		return fmt.Errorf("can't make change: current validator weight %d exceeds max allowed weight change of %d", vdrInfo.Weight, uint64(allowedChange))
+	}
+
+	allowedChange = float64(totalWeight-vdrInfo.Weight) * constants.MaxL1TotalWeightChange
+
+	if newWeight == 0 {
+		ux.Logger.PrintToUser("Current validator weight is %d", vdrInfo.Weight)
+		newWeight, err = app.Prompt.CaptureWeight(
+			"What weight would you like to assign to the validator?",
+			func(v uint64) error {
+				if v > uint64(allowedChange) {
+					return fmt.Errorf("weight exceeds max allowed weight change of %d", uint64(allowedChange))
+				}
+				return nil
+			},
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	if float64(newWeight) > allowedChange {
+		return fmt.Errorf("can't make change: desired validator weight %d exceeds max allowed weight change of %d", newWeight, uint64(allowedChange))
+	}
+
+	publicKey, err = formatting.Encode(formatting.HexNC, bls.PublicKeyToCompressedBytes(vdrInfo.PublicKey))
 	if err != nil {
 		return err
 	}
 
-	remainingBalanceOwnerAddr, err = getChangeAddrFromPChain()
-	if err != nil {
-		return fmt.Errorf("failure parsing change owner address: %w", err)
+	if pop == "" {
+		_, pop, err = promptProofOfPossession(false, true)
+		if err != nil {
+			return err
+		}
 	}
 
-	// add back validator to subnet with updated weight
-	return CallAddValidator(deployer, network, kc, blockchainName, nodeID.String(), publicKey, pop)
-}
+	deployer := subnet.NewPublicDeployer(app, kc, network)
 
-// getValidatorBalanceFromPChain gets remaining balance of validator from p chain
-func getValidatorBalanceFromPChain() (uint64, error) {
-	return 0, nil
-}
+	var remainingBalanceOwnerAddr, disableOwnerAddr string
+	hrp := key.GetHRP(network.ID)
+	if len(vdrInfo.RemainingBalanceOwner.Addrs) > 0 {
+		remainingBalanceOwnerAddr, err = address.Format("P", hrp, vdrInfo.RemainingBalanceOwner.Addrs[0][:])
+		if err != nil {
+			return err
+		}
+	}
+	if len(vdrInfo.DeactivationOwner.Addrs) > 0 {
+		disableOwnerAddr, err = address.Format("P", hrp, vdrInfo.DeactivationOwner.Addrs[0][:])
+		if err != nil {
+			return err
+		}
+	}
 
-// getBLSInfoFromPChain gets BLS public key and pop from info api
-func getBLSInfoFromPChain() (string, string, error) {
-	return "", "", nil
-}
+	// first remove the validator from subnet
+	err = removeValidatorSOV(
+		deployer,
+		network,
+		blockchainName,
+		nodeID,
+		0, // automatic uptime
+		isBootstrapValidatorForNetwork(nodeID, sc.Networks[network.Name()]),
+		false, // don't force
+	)
+	if err != nil {
+		return err
+	}
 
-// getChangeAddrFromPChain gets validator change addr from info api
-func getChangeAddrFromPChain() (string, error) {
-	return "", nil
+	balance := vdrInfo.Balance
+	if len(vdrInfo.RemainingBalanceOwner.Addrs) > 0 {
+		availableBalance, err := utils.GetNetworkBalance([]ids.ShortID{vdrInfo.RemainingBalanceOwner.Addrs[0]}, network.Endpoint)
+		if err != nil {
```

Comment:
I am concerned that if  we change weight by removing and adding validator back we should not stop and return err in the middle, we should try to recover and add validator back at least with the same weight.

maybe it makes sense to have error passed to `defer func` that can try adding a validator back if `err` is not nil